File: leetcode/solved/24_Swap_Nodes_in_Pairs/solution.cpp
/*
 * solution.cpp
 * Copyright (C) 2018 sungmin <smoh2044@gmail.com>
 *
 * Distributed under terms of the MIT license.
 */

#include <iostream>
#include <vector>
#include <string>
#include <sstream>


using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(head == NULL) return head;
        ListNode* p = NULL;
        ListNode* l = head;
        ListNode* r = head->next;
        if(r == NULL) return head;
        l->next = r->next;
        r->next = l;
        ListNode* newHead = r;
        r = l;
        while(r->next && r->next->next){
            p = r;
            l = r->next;
            r = r->next->next;
            p->next = r;
            l->next = r->next;
            r->next = l;
            r = l;
        }
        return newHead;
    }
};


ListNode* getInput(){
    string s;
    getline(cin, s);
    stringstream ss(s);
    int n;
    ss >> n;
    ListNode* head = new ListNode(n);
    ListNode* p = head;
    while(ss >> n){
        ListNode* node = new ListNode(n);
        p->next = node;
        p = p->next;
    }
    return head;
}

void printList(ListNode* head){
    ListNode* p = head;
    while(p){
        cout << p->val << " ";
        p = p->next;
    }
    cout << endl;
}

int main(){
    ListNode* head = getInput();
    printList(Solution().swapPairs(head));
    return 0;
}

File: leetcode/solved/24_Swap_Nodes_in_Pairs/solution.py
from typing import Optional

#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2020 sungminoh <smoh2044@gmail.com>
#
# Distributed under terms of the MIT license.

"""
Given a linked list, swap every two adjacent nodes and return its head. You must solve the problem without modifying the values in the list's nodes (i.e., only nodes themselves may be changed.)

Example 1:

Input: head = [1,2,3,4]
Output: [2,1,4,3]

Example 2:

Input: head = []
Output: []

Example 3:

Input: head = [1]
Output: [1]

Constraints:

	The number of nodes in the list is in the range [0, 100].
	0 <= Node.val <= 100
"""
import pytest
import sys
sys.path.append('../')
from exercise.list import ListNode, build_list


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """Feb 28, 2022 12:18"""
        if not head:
            return
        if not head.next:
            return head
        nodes = [head, head.next]
        nodes[0].next = self.swapPairs(nodes[1].next)
        nodes[1].next = nodes[0]
        return nodes[1]

    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """Sep 22, 2023 15:48"""
        if not head:
            return head

        def swap(p, n, nn):
            p.next, n.next, nn.next = nn, nn.next, n

        dummy = ListNode()
        dummy.next = head
        p = dummy
        n = head
        while n and n.next:
            swap(p, n, n.next)
            p, n = n, n.next
        return dummy.next


@pytest.mark.parametrize('args', [
    (([1,2,3,4], [2,1,4,3])),
    (([], [])),
    (([1], [1])),
])
def test(args):
    assert build_list(args[-1]) == Solution().swapPairs(build_list(args[0]))


if __name__ == '__main__':
    sys.exit(pytest.main(["-s", "-v"] + sys.argv))
